### PR TITLE
feat: DID package refactor

### PR DESCRIPTION
**Pull request by Diiaablo95**
# DID package refactor

This PR fixes https://github.com/KILTprotocol/ticket/issues/1756 and fixes https://github.com/KILTprotocol/ticket/issues/1712 by making DIDs and everything that uses DIDs (e.g., extrinsic DID signing) easier to use. The spec doc can be found in [this HackMD document](https://hackmd.io/q2Sh5u6LTk-2WG2FCPvMIg). Most of the things match the implementation, others have deviated a bit during the development.

## What's changed

Most of the design was specified in the design document above, with the main different compared to the current codebase being that there are two components involved for DIDs:

- a `DidDetails` abstract class which is then derive by `LightDidDetails` and `FullDidDetails`. These classes are meant to be used on the owner side (e.g., for decrypting/signing stuff, or to sign extrinsics).
- a `IDidDetails` interface which is implemented by both `LightDidDetails` and `FullDidDetails`. This interface is meant to be used on the receiver side, hence it is exposed by the resolver and by methods that take arguments referring to the "other communicating party" (e.g., during credential exchange). This interface exposes basic methods that one would expect from a DID, i.e., retrieving key(s), and service endpoints by ID or by type.

## Other changes

Some other important changes are the following:

- rename `DefaultResolver` to `DidResolver`
- refactor utility functions used in the integration tests to hide most of the logic from common operations and initialisation code (quite some more improvements could be done here, e.g. in [#1759](https://github.com/KILTprotocol/ticket/issues/1759).
- changes to how messages/request for attestations/presentations are authenticated with DIDs

### Batched DID operations (for creations and updates)

This PR has not implemented a DID builder, which would make the generation of full DID way easier. The initial design can be found in the document linked above, while a almost complete (untested) implementation can be found in this PR history, specifically commit [1cdc6e](https://github.com/KILTprotocol/sdk-js/pull/447/commits/1cdc6eb241fa4feddaa8805daa34a9721274f67d) is the one where the whole code for builders was removed since it was decided to put it in a different PR.

### Updated getting-started and examples guides

Another thing that is left out of this PR is the documentations. The reason is that most of the docs will have to change again fairly shortly when batched operations are added, hence I though that it would not be the best use of time to change all docs now and then change them again in few weeks. We won't have an official release that external people will be using anyway before we add the batching, so I don't think it's a big problem if integration tests for the docs are failing for a while. There's anyway a [ticket](https://github.com/KILTprotocol/ticket/issues/1760) tracking that.

### CODE REVIEW

File: packages/core/src/__integrationtests__/Deposit.spec.ts [hunk 1]
```diff
@@ -9,111 +9,87 @@
...
 import { BN } from '@polkadot/util'
 import {
-  createMinimalFullDidFromLightDid,
-  WS_ADDRESS,
+  createFullDidFromLightDid,
+  createFullDidFromSeed,
+  createMinimalLightDidFromSeed,
   devFaucet,
-  DriversLicense,
+  driversLicenseCTypeForDeposit as driversLicenseCType,
```
Comment (Diiaablo95) [inline packages/core/src/__integrationtests__/Deposit.spec.ts:27]: Apparently there was a concurrency issue only when the tests are running in the CI ("it works on my machine"), for which I had to create a separate CType and use that throughout the deposit claiming tests.

**Review by Dudleyneedham — CHANGES_REQUESTED** [review]
At the moment, it feels simpler. Though I would need to use it and go through the workflow in order to see it all works. 

Minor concern about the readability with the syntax chaining.  The `DidDetails` `getKeys` is an example of this. Its hard to follow. 

Overall, looking for to having this in and testing it. Looks good so far.

File: packages/did/src/Did.chain.ts [hunk 2]
```diff
@@ -5,58 +5,110 @@
...
-  encodeServiceEndpoint,
-  parseDidUrl,
-  assembleDidFragment,
-} from './Did.utils.js'
+
+import { DidDetails, getSignatureAlgForKeyType } from './DidDetails/index.js'
+
+// ### Chain type definitions
+
+type KeyId = Hash
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts:49]: No, these are all internal types for raw querying. The result of the raw query is then wrapped in a nicer encoded format for users to use.

File: packages/did/src/Did.chain.ts [hunk 3]
```diff
@@ -5,58 +5,110 @@
...
+}
 
 // ### RAW QUERYING (lowest layer)
 
 // Query a full DID given the identifier (a KILT address for v1).
 // Interacts with the Did storage map.
-export async function queryDidEncoded(
-  didIdentifier: IIdentity['address']
+async function queryDidEncoded(
+  didIdentifier: IDidIdentifier
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts:104]: Yes. This is a chain file, and query on chain is performed using only identifiers and not full DIDs. Hence querying details, endpoint counts, deletion status, etc, all takes place via a DID identifier. There's also an `identifier` getter in  the `DidDetails` class to make it simpler to access such property.

File: packages/did/src/Did.chain.ts [hunk 4]
```diff
@@ -106,248 +144,277 @@ export async function queryAllServicesEncoded(
...
-  const publicKeys: IDidKeyDetails[] = Array.from(
-    didDetail.publicKeys.entries()
-  ).map(([keyId, keyDetails]) => {
-    return decodeDidPublicKeyDetails(did, keyId, keyDetails)
-  })
-  const authenticationKeyId = assembleKeyId(didDetail.authenticationKey, did)
-  const keyAgreementKeyIds = Array.from(
-    didDetail.keyAgreementKeys.values()
-  ).map((id) => assembleKeyId(id, did))
+  const publicKeys: DidKey[] = [...didDetail.publicKeys.entries()].map(
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts:201]: But we know it is, because `IDidChainRecordCodec` extends `Struct`, and if this is not of the proper type, the function would just fail to decode. Am I missing something here? Or what would you suggest changing here?

File: packages/did/src/Did.chain.ts [hunk 5]
```diff
@@ -106,248 +144,277 @@ export async function queryAllServicesEncoded(
...
+    ([keyId, keyDetails]) => {
+      return decodeDidPublicKeyDetails(keyId, keyDetails)
+    }
+  )
+  const authenticationKeyId = didDetail.authenticationKey.toHex()
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts:206]: Authentication key can be rotated, but it is the only key that is required to always be present. Hence, it can never be undefined, otherwise the DID would become unusable.

File: packages/did/src/Did.chain.ts [hunk 6]
```diff
@@ -106,248 +144,277 @@ export async function queryAllServicesEncoded(
...
 }
 
+export async function queryDidDeletionStatus(
+  didIdentifier: IDidIdentifier
+): Promise<boolean> {
+  const { api } = await BlockchainApiConnection.getConnectionOrConnect()
+  const encodedStorageHash = await api.query.did.didBlacklist.hash(
+    didIdentifier
+  )
+  return !encodedStorageHash.isEmpty
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts:299]: If `isEmpty` is true, it means that there is no entry in the set of deleted DIDs, so the function should return `false`. If `isEmpty` is false, the function should return `true`. Thus the negation.

File: packages/did/src/DidDetails/DidDetails.ts [hunk 7]
```diff
@@ -5,82 +5,159 @@
...
+      ...endpointDetails,
     }
-    return [...this.keys.keys()]
   }
 
-  getEndpointById(id: string): IDidServiceEndpoint | undefined {
-    return this.services.get(id)
+  public getEndpoints(type?: string): DidServiceEndpoint[] {
+    const serviceEndpointsEntries = type
+      ? [...this.serviceEndpoints.entries()].filter(([, details]) => {
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/DidDetails/DidDetails.ts:123]: There's a lint warning generated if I add that:

`'_' is defined but never used.`

File: packages/did/src/DidDetails/FullDidDetails.ts [hunk 8]
```diff
@@ -6,178 +6,150 @@
...
-   * @returns The set of keys that could be used to sign the extrinsic.
-   */
-  public getKeysForExtrinsic(
-    apiOrMetadata: ApiOrMetadata,
-    extrinsic: Extrinsic
-  ): IDidKeyDetails[] {
-    return getKeysForExtrinsic(apiOrMetadata, this, extrinsic)
+  async getNextNonce(): Promise<BN> {
+    const currentNonce = await queryNonce(this.identifier)
+    // Wrap around the max u64 value when reached.
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/DidDetails/FullDidDetails.ts:122]: Well, no, but if it does...? 😂

File: packages/did/src/DidDetails/LightDidDetails.ts [hunk 9]
```diff
@@ -39,56 +78,125 @@ export class LightDidDetails extends DidDetails {
...
+    }
+    return LightDidDetails.fromDetails({
+      authenticationKey,
+    })
+  }
+
+  public async migrate(
+    submitterAddress: IIdentity['address'],
+    signer: KeystoreSigner,
+    migrationHandler: DidMigrationHandler
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/DidDetails/LightDidDetails.ts:184]: That will be dealt with in the `DidBuilder` class, as specified in https://hackmd.io/q2Sh5u6LTk-2WG2FCPvMIg#FullDidBuilder-class-not-yet-implemented-subject-to-changes. So, this `migrate` would simply take whatever state the light DID has, and migrate it on chain. For more complex operation, the builder should be used (when we will implement it).

File: packages/did/src/DidDetails/DidDetails.utils.ts [hunk 10]
```diff
@@ -5,103 +5,59 @@
...
-    alg: getSignatureAlgForKeyType(signingKey.type),
-    didIdentifier: getIdentifierFromKiltDid(didDetails.did),
-    submitter,
-    keys,
-    endpoints: didDetails.getEndpoints(),
+  const allowedKeyRelationships: Set<string> = new Set([
+    ...Object.values(KeyRelationship),
+    'none',
+  ])
+  Object.keys(keyRelationships).forEach((kr) => {
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/DidDetails/DidDetails.utils.ts]: Good point. Fixed in https://github.com/KILTprotocol/sdk-js/pull/447/commits/15350115a3aaf04f51e7ad752060c44a8aad13e4

File: packages/did/src/DidDetails/DidDetails.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/did/src/DidDetails/DidDetails.ts:123]: makes it more explicit, but also assigns an unused variable, which may introduce a pointless performance penalty? not sure how it works though

File: packages/did/src/Did.chain.ts [hunk 11]
```diff
@@ -411,32 +478,38 @@ export async function getAddKeyExtrinsic(
...
-  const encoded = encodeServiceEndpoint(api.registry, endpoint)
-  return api.tx.did.addServiceEndpoint(encoded)
+  const encodedEndpoint =
+    new (api.registry.getOrThrow<IServiceEndpointChainRecordCodec>(
+      'DidServiceEndpointsDidEndpoint'
+    ))(api.registry, {
+      serviceTypes: endpoint.types,
+      ...endpoint,
+    })
+  return api.tx.did.addServiceEndpoint(encodedEndpoint)
```
Comment (rflechtner) [inline packages/did/src/Did.chain.ts:491]: this should suffice:
```suggestion
    return api.tx.did.addServiceEndpoint( {
      serviceTypes: endpoint.types,
      ...endpoint,
    })
```
we only needed to encode explicitly in cases where we had to sign the encoded representation

**Review by rflechtner — APPROVED** [review]
There is so much new stuff for me since I last touched all this that I'm not the best to judge the changes introduced here. But looks quite nice to my eyes. I've left a few comments and suggestions regarding details.

File: packages/did/src/Did.chain.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply packages/did/src/Did.chain.ts:49]: Ok, good to know.

File: packages/did/src/Did.chain.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply packages/did/src/Did.chain.ts:104]: Ok, that is good to know.

File: packages/did/src/Did.chain.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply packages/did/src/Did.chain.ts:299]: A comment explaining should help out.

File: packages/did/src/DidDetails/DidDetails.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply packages/did/src/DidDetails/DidDetails.ts:123]: Ok, no need. It would read a bit better.

File: packages/did/src/DidDetails/FullDidDetails.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply packages/did/src/DidDetails/FullDidDetails.ts:122]: It would be terrifying to think.

File: packages/did/src/DidDetails/LightDidDetails.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply packages/did/src/DidDetails/LightDidDetails.ts:184]: Ok, good to know.

File: packages/did/src/Did.chain.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts:301]: https://github.com/KILTprotocol/sdk-js/pull/447/commits/e3164f5d88e7b8c8049ba5c025ce751bf5ad9c3e

File: packages/did/src/DidDetails/FullDidDetails.ts [hunk 12]
```diff
@@ -6,178 +6,150 @@
...
-   * @param apiOrMetadata The node runtime information to use to retrieve the required information.
-   * @param extrinsic The extrinsic to submit.
-   * @returns The set of keys that could be used to sign the extrinsic.
-   */
-  public getKeysForExtrinsic(
-    apiOrMetadata: ApiOrMetadata,
-    extrinsic: Extrinsic
-  ): IDidKeyDetails[] {
-    return getKeysForExtrinsic(apiOrMetadata, this, extrinsic)
+  async getNextNonce(): Promise<BN> {
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/DidDetails/FullDidDetails.ts]: https://github.com/KILTprotocol/sdk-js/pull/447/commits/f8cc40244ab7c1a96c737ed3fa4aeefd8ddc5abf

File: packages/did/src/DidDetails/FullDidDetails.ts [hunk 13]
```diff
@@ -6,178 +6,150 @@
...
-    signer: KeystoreSigner,
-    submitterAccount: IIdentity['address'],
-    incrementTxIndex = true
+    {
+      signer,
+      submitterAccount,
+      keySelection = defaultDidKeySelection,
+    }: {
+      signer: KeystoreSigner
+      submitterAccount: IDidIdentifier
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/DidDetails/FullDidDetails.ts]: Good point -> https://github.com/KILTprotocol/sdk-js/pull/447/commits/3596ef32dd5304c3385786ca84d4412c64b6b109

File: packages/did/src/Did.utils.ts [hunk 14]
```diff
@@ -202,11 +154,11 @@ export function validateKiltDid(
   return true
 }
 
-export function validateDidSignature(input: unknown): input is DidSignature {
+export function validateDidSignature(signature: DidSignature): boolean {
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.utils.ts]: https://github.com/KILTprotocol/sdk-js/pull/447/commits/17cecacd7bf6b9db08d5e30d34e2b9db168a2181

File: packages/did/src/Did.chain.ts [hunk 15]
```diff
@@ -5,58 +5,110 @@
...
-  didIdentifier: IIdentity['address']
+async function queryDidEncoded(
+  didIdentifier: IDidIdentifier
 ): Promise<Option<IDidChainRecordCodec>> {
   const { api } = await BlockchainApiConnection.getConnectionOrConnect()
   return api.query.did.did<Option<IDidChainRecordCodec>>(didIdentifier)
 }
 
 // Query ALL deleted DIDs, which can be very time consuming if the number of deleted DIDs gets large.
-export async function queryDeletedDidsEncoded(): Promise<GenericAccountId[]> {
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts:59]: We can maybe do these sort of changes in a different PR, I would say.

File: packages/did/src/Did.chain.ts [hunk 16]
```diff
@@ -106,248 +144,277 @@ export async function queryAllServicesEncoded(
...
+      .pop() || undefined
+
+  // For now, it only takes the first delegation key, if present.
+  const newDelegationKey: PublicKeyEnum | undefined =
+    did
+      .getKeys(KeyRelationship.capabilityDelegation)
+      .map((key) => {
+        return formatPublicKey(key)
+      })
+      .pop() || undefined
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts]: https://github.com/KILTprotocol/sdk-js/pull/447/commits/eebefffa40b143b15564e9b680733624586926cf

File: packages/did/src/Did.chain.ts [hunk 17]
```diff
@@ -106,248 +144,277 @@ export async function queryAllServicesEncoded(
...
+    newKeyAgreementKeys,
+    newAttestationKey,
+    newDelegationKey,
+    newServiceDetails,
+  }
+
+  const encodedDidCreationDetails =
+    new (api.registry.getOrThrow<IDidCreationDetails>(
+      'DidDidDetailsDidCreationDetails'
+    ))(api.registry, rawCreationDetails)
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts]: https://github.com/KILTprotocol/sdk-js/pull/447/commits/f324954ce72d53bc357955cb802ebfd65297799f

File: packages/did/src/Did.chain.ts (hunk 11 above)
Comment (Diiaablo95) [reply packages/did/src/Did.chain.ts]: https://github.com/KILTprotocol/sdk-js/pull/447/commits/f324954ce72d53bc357955cb802ebfd65297799f

**Review by Dudleyneedham — APPROVED** [review]
LGTM

File: packages/did/src/DidDocumentExporter/README.md [hunk 18]
```diff
@@ -1,52 +0,0 @@
-# DID Document exporter
```
Comment (tjwelde) [inline packages/did/src/DidDocumentExporter/README.md:1]: Q: Did this documentation move somewhere else?

File: packages/did/src/DidDetails/LightDidDetails.utils.ts [hunk 19]
```diff
@@ -107,3 +155,14 @@ export function decodeAndDeserializeAdditionalLightDidDetails(
     serviceEndpoints: deserialized[SERVICES_KEY_MAP_KEY],
   }
 }
+
+export function getDefaultMigrationHandler(
```
Comment (tjwelde) [inline packages/did/src/DidDetails/LightDidDetails.utils.ts:159]: Q: Should this maybe be moved into the integrationtest utils? Seems to be only used there and I wouldn't expose a function which does resigning per default to the user of the sdk

File: packages/did/src/DidDetails/FullDidDetails.ts [hunk 20]
```diff
@@ -6,178 +6,191 @@
...
 
-  /**
-   * Signs and returns the provided unsigned extrinsic with the right DID key, if present. Otherwise, it will return an error.
-   *
-   * @param extrinsic The unsigned extrinsic to sign.
-   * @param signer The keystore to be used to sign the encoded extrinsic.
-   * @param submitterAccount The KILT account to bind the DID operation to (to avoid MitM and replay attacks).
-   * @param incrementTxIndex Flag indicating whether the DID nonce should be increased before submitting the operation or not.
-   * @returns The DID-signed submittable extrinsic.
-   */
```
Comment (tjwelde) [inline packages/did/src/DidDetails/FullDidDetails.ts:149]: Is this jsdoc comment already on the interface? If not I would say that these are quite helpful and should be added again.

**Review by tjwelde — COMMENTED** [review]
I really like it. Just some questions and comments because of consistent coding style

File: packages/core/src/quote/Quote.ts [hunk 21]
```diff
@@ -99,17 +104,36 @@ export async function fromAttesterSignedInput(
 
 export async function createAttesterSignature(
   quoteInput: IQuote,
-  attesterIdentity: IDidDetails,
-  signer: KeystoreSigner
+  attesterIdentity: DidDetails,
+  {
+    signer,
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/core/src/quote/Quote.ts:109]: Yes, the idea is that anything that is domain-specific, like the quote input and the attester, are parameters, and everything that is the same across several operations, like the keystore, the key selection, the resolver in some cases, would be in an object, as to make it more memorable for users to interact with. What do you think? Otherwise we would end up having four to six parameters, whose order might be hard to remember.

File: packages/core/src/credential/Credential.ts [hunk 22]
```diff
@@ -249,15 +249,20 @@ export class Credential implements ICredential {
...
 
-    if (claimerDid) {
-      await credential.request.signWithDid(signer, claimerDid, challenge)
-    } else if (claimerSigningKey) {
-      await credential.request.signWithKey(signer, claimerSigningKey, challenge)
-    } else {
-      throw new Error(
-        'Either a key or claimer did details are required for signing'
-      )
+    const keys = claimerDid.getKeys(KeyRelationship.authentication)
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/core/src/credential/Credential.ts:252]: Where? We don't take keys anymore, but just the claimer `DidDetails` and the key selection closure.

[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/core/src/credential/Credential.ts:252]: But if your question was wether it is ok, I think it is, for the reason that you mentioned. For presentation we always use authentication keys, so providing a closure to decide which key to select replaces the fact of having to manually pass in the key. We just fetch them from the given details and let the user choose which one to use.

File: packages/messaging/src/Message.ts [hunk 23]
```diff
@@ -97,47 +97,42 @@ export class Message implements IMessage {
...
-      resolver = DefaultResolver,
+      resolver = DidResolver,
     }: {
-      senderDetails?: IDidDetails
-      receiverDetails?: IDidDetails
+      keystore: Pick<NaclBoxCapable, 'decrypt'>
+      receiverDetails: DidDetails
       resolver?: IDidResolver
-    } = {}
+    }
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/messaging/src/Message.ts:126]: In this case `keystore` and `receiverDetails` are not optional. Only the resolver is. Plus, it's for consistency as mentioned before, but I'm definitely open to alternative ways of doing it. What would be your suggestion?

File: packages/did/src/DidDocumentExporter/README.md (hunk 18 above)
Comment (Diiaablo95) [reply packages/did/src/DidDocumentExporter/README.md]: Was lost in one of the merges -> https://github.com/KILTprotocol/sdk-js/pull/447/commits/70f0cd2fd85255e6a289ee1f7a954d154f854208

File: packages/did/src/DidDetails/LightDidDetails.utils.ts (hunk 19 above)
Comment (Diiaablo95) [reply packages/did/src/DidDetails/LightDidDetails.utils.ts]: Makes sense -> https://github.com/KILTprotocol/sdk-js/pull/447/commits/d669855f5582a3e97384bf86fe8d1a1240876a66

File: packages/did/src/DidDetails/FullDidDetails.ts (hunk 20 above)
Comment (Diiaablo95) [reply packages/did/src/DidDetails/FullDidDetails.ts:149]: Yes, I never added them again, so thanks for the heads up -> https://github.com/KILTprotocol/sdk-js/pull/447/commits/717e627262f495ebc47334416e6893bcc1243d80

File: packages/messaging/src/Message.ts [hunk 24]
```diff
@@ -97,47 +97,42 @@ export class Message implements IMessage {
...
+    if (!senderKeyDetails) {
+      throw new Error(`Could not resolve sender key ${senderKeyId}`)
+    }
+    const { fragment } = DidUtils.parseDidUri(receiverKeyId)
+    if (!fragment) {
+      throw new Error(`No fragment for the receiver key ID ${receiverKeyId}`)
+    }
+    const receiverKeyDetails = receiverDetails.getKey(fragment)
+    if (!receiverKeyDetails) {
+      throw new Error(`Could not resolve receiver key ${receiverKeyId}`)
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply packages/messaging/src/Message.ts]: Is this suitable? https://github.com/KILTprotocol/sdk-js/pull/447/commits/46efcec2db3de7e3fcdb09f5b08913db4361997a

Or do we want to use a different error for all potential errors?

File: packages/core/src/quote/Quote.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (tjwelde) [reply packages/core/src/quote/Quote.ts:109]: as discussed, let's just put optional parameters into the object and also assign that to the empty object.

File: packages/messaging/src/Message.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (tjwelde) [reply packages/messaging/src/Message.ts:126]: same as above

File: packages/did/src/Did.chain.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/did/src/Did.chain.ts:59]: fine by me
